---
title: "Merging"
output: 
  flexdashboard::flex_dashboard:
    orientation: rows
    vertical_layout: scroll
    theme: sandstone
---


```{r, include=FALSE,message=F}
suppressPackageStartupMessages(require(data.table))
  merge_untrimmed.file <- snakemake@input[["mergedata_untrimmed"]]
  merge_trimmed.file <- snakemake@input[["mergedata_trimmed"]]
  merge_inserts.file <- snakemake@input[["mergdata_trimmed_details"]]
  
  sampleoverview_merge.file <- snakemake@input[["merging_trimmed_overview"]]
  tplot_merge.file <- snakemake@input[["merging_untrimmed_toplot"]]
  
  sampleoverview_merge_untr.file <- snakemake@input[["merging_untrimmed_overview"]]
  tplot_merge_untr.file <- snakemake@input[["merging_untrimmed_toplot"]]
  
  threads = snakemake@threads
  commonscript.file = snakemake@input[["commons"]]
  
  fastqcreports.tr = as.data.table(snakemake@params[["merging_trimmed_fastqc"]])
  fastqcreports.tr = fastqcreports.tr[,qcreport:=snakemake@input[["merging_trimmed_fastqc"]]]
  fastqcreports.utr = as.data.table(snakemake@params[["merging_untrimmed_fastqc"]])
  fastqcreports.utr = fastqcreports.utr[,qcreport:=snakemake@input[["merging_untrimmed_fastqc"]]]
  
files = c(merge_untrimmed.file,merge_trimmed.file,merge_inserts.file,sampleoverview_merge.file,tplot_merge.file,sampleoverview_merge_untr.file,tplot_merge_untr.file)

source(commonscript.file)
commonOptions(threads)
suppressPackageStartupMessages(require(matrixStats))
samplenames <- list(sample=snakemake@params[["samples"]])


merge_untrimmed<-readData(merge_untrimmed.file, samplenames)
merge_trimmed<-readData(merge_trimmed.file, samplenames)



merge_inserts<-readData(merge_inserts.file, samplenames, readfun=function(...)fread(..., sep="\t",fill=T))
merge_inserts<-merge_inserts[insert!=-1,density(insert)[c("x","y")],by="sample"]

sampleoverview_merge <- readData(sampleoverview_merge.file, samplenames)
sampleoverview_merge <- fastqcreports.tr[sampleoverview_merge,on=c("sample","read")]
tplot_merge <- readData(tplot_merge.file, samplenames)


sampleoverview_merge_untr <- readData(sampleoverview_merge_untr.file, samplenames)
sampleoverview_merge_untr = fastqcreports.utr[sampleoverview_merge_untr,on=c("sample","read")]
tplot_merge_untr <- readData(tplot_merge_untr.file, samplenames)

sampleoverview_merge[,trimmed:="Trimmed"]
sampleoverview_merge_untr[,trimmed:="Untrimmed"]
sampleoverview = rbind(sampleoverview_merge,sampleoverview_merge_untr)

tplot_merge[,trimmed:="Trimmed"]
tplot_merge_untr[,trimmed:="Untrimmed"]
tplot = rbind(tplot_merge,tplot_merge_untr)
tplot[,read:=paste0(trimmed,".",read)]

```


Overview
==================

Row
--------------------------------------

### Application Note

For comparison, the merging step has been done with (adapter-)trimmed and untrimmed sequences. 

> Rendered at: `r Sys.time()`.

Row
--------------------------------------

### File Note

```{r}
knitr::kable(genInfoBlock(files))
```

> The table shows the files which have been used to generate this report.


Row
--------------------------------------

### Merging results

```{r}
mergestats <- c("joined","ambiguous","nosolution","tooshort")
merge_untrimmed[,id:=paste0(sample,".untrimmed")]
merge_trimmed[,id:=paste0(sample,".trimmed")]
merge = rbind(merge_trimmed,merge_untrimmed)
merge[,(mergestats):=lapply(.SD,function(col)col/pairstomerge*100),.SD=mergestats]
merge.molten<-melt(merge,id.vars = "id",measure.vars = mergestats)

ggplotly(ggplot(merge.molten, aes(x=id, y=value,fill=variable,group=id)) + 
  geom_bar(stat = "identity") + xlab("Sample") + ylab("Percentage of read pairs") +  scale_y_continuous(labels = function(x) paste0(x, "%")) + scale_fill_brewer(palette = additional_palette) +labs(fill="Merge Result"))
```

>This plot shows the result statistics for the pairs during the BBmerge run. The trimmed sequences were used in the following workflow steps. The [algorithm](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0185056#sec006) has different reasons for rejecting a pairing with "no solution". "Ambiguous" means that the two best (lowest) (matches+constant)/(mismatches+matches) ratios were too similar or that the second best alignment ratio was below a cutoff. "Too short" are reads whose number of matches in the alignment was lower than you would expect for the complexity of the given pair. 

### Insert sizes

```{r}
ggplotly(ggplot(merge_inserts) + aes(x=x,y=y,color=sample) +geom_line() + scale_fill_brewer(palette = sample_palette) + xlab("Insert(Overlap) length in bp") + ylab("Density")+labs(color="Sample"))
```

>In this plot the distribution of the insert (overlap) length in the different samples is shown with the (estimated) density.

Row
--------------------------------------

### Merging results

```{r}
sampleoverview.vis = sampleoverview[,.(sample,trimmed,read,file,qcreport,minReadLen,maxReadLen,reads)]
colnam = c(sample="Sample",trimmed="Trimmed",read="Group",qcreport="FASTQC Report",minReadLen="Shortest Read [bp]",maxReadLen="Longest Read [bp]",reads="Number of read pairs")
setnames(sampleoverview.vis,names(colnam),colnam)
datatable(sampleoverview.vis,escape = F)
```

>This table shows the files resulting from the merge.

Row
--------------------------------------

### Read lengths after merging

```{r}
ggplot(sampleoverview) + aes(x=read,lower=quantile25ReadLen,upper=quantile75ReadLen,middle=medianReadLen,ymin=minReadLen,ymax=maxReadLen,group=interaction(sample,trimmed,read),fill=sample) + geom_boxplot(stat="identity",position = "dodge")+scale_fill_brewer(palette = sample_palette) + xlab ("Read group") +ylab("Read length in bp") + facet_wrap(~trimmed) + labs(fill="Sample") + theme(legend.position = "bottom")
```

> The distribution of the read lengths within each sample for each merging operation as a boxplot, the whiskers extend to the min/max values.


### Read length

```{r}
dat = tplot[type=="readlength_density" & x>=min(sampleoverview$minReadLen) & x<=max(sampleoverview$maxReadLen),]
#dat = tplot[type=="gccontent_density" & x>=0 & x<=1,]
ggplotly(ggplot(dat) + aes(x=x,y=y,linetype=read,color=sample)  + geom_line() +scale_color_brewer(palette = sample_palette)+ xlab("Read Length in bp") + facet_wrap(~trimmed) +  ylab("Density Estimate") +theme(legend.title = element_blank())) 
```

> The read length in each sample and merge result. The plot has been trimmed at the actual min and max read lengths.


Row
--------------------------------------

### Sequence Quality Scores

```{r}
dat = tplot[type=="avg_seq_qual_dens",]
#dat = tplot[type=="gccontent_density" & x>=0 & x<=1,]
ggplotly(ggplot(dat) + aes(x=x,y=y,linetype=read,color=sample)  + geom_line() +scale_color_brewer(palette = sample_palette)+ xlab("Average Phred Score for a whole read") + ylab("Density Estimate") +theme(legend.title = element_blank()) + facet_wrap(~trimmed))
```

> The (estimated) density distribution of the average Phred score for the sequences. This plot has not been trimmed!


### Cycle Quality Scores

```{r}
dat = tplot[type=="cycle_qual_counts",.(median=weightedMedian(y,z)),by=c("sample","read","x","trimmed")]
ggplotly(ggplot(dat) + aes(x=x,y=median,color=sample,linetype=read) +geom_line()+scale_color_brewer(palette = sample_palette)+ ylab( "Median phred score")+xlab("Cycle")+theme(legend.title = element_blank())+ facet_wrap(~trimmed))
```

> The median of the phred score at the given cycle/position for a sample and read type
