---
title: "Read trimming"
output: 
  flexdashboard::flex_dashboard:
    orientation: rows
    vertical_layout: scroll
    theme: sandstone
---


```{r, include=FALSE}
suppressPackageStartupMessages(require(data.table))

merge_untrimmed.file <- snakemake@input[["mergedata_untrimmed"]]
merge_trimmed.file <- snakemake@input[["mergedata_trimmed"]]
trim_summary.file <- snakemake@input[["trimming_summary"]]
trim_details.file <- snakemake@input[["trimming_details"]]
trim_comparison.file <- snakemake@input[["trim_summary_impact"]]
sampleoverview_trim.file <- snakemake@input[["trimmed_overview"]]  
tplot_raw.file <- snakemake@input[["raw_toplot"]]
sampleoverview_raw.file <- snakemake@input[["raw_overview"]]
tplot_trimmed.file <- snakemake@input[["trimmed_toplot"]]

conf_adapters <- snakemake@params[["adapters"]]

threads <- snakemake@threads
commonscript.file <- snakemake@input[["commons"]]

fastqcreports <- cbind(as.data.table(snakemake@params$samples_reads), qcreport=snakemake@input[["fastqc"]])

files <- unlist(c(merge_untrimmed.file,merge_trimmed.file,trim_summary.file,trim_details.file,trim_comparison.file,sampleoverview_trim.file,tplot_raw.file,tplot_trimmed.file))

source(commonscript.file)
commonOptions(threads)
suppressPackageStartupMessages(require(matrixStats))
samplenames <- list(sample=snakemake@params[["samples"]])

merge_untrimmed<-readData(merge_untrimmed.file, samplenames)
merge_untrimmed[,adaptercount:=round(adaptercount/pairstomerge*100,2)]
merge_trimmed<-readData(merge_trimmed.file, samplenames)
merge_trimmed[,adaptercount:=round(adaptercount/pairstomerge*100,2)]

trim_comparison <- readData(trim_comparison.file, samplenames)

sampleoverview_trim <- readData(sampleoverview_trim.file, samplenames)
sampleoverview_raw <- readData(sampleoverview_raw.file, samplenames)

tplot_raw <- readData(tplot_raw.file, samplenames)
tplot_trimmed <- readData(tplot_raw.file, samplenames)

tplot <- rbind(tplot_raw[type=="cycle_qual_counts",.(trimmed="trimmed",median=weightedMedian(y,z)),by=c("sample","read","x")],tplot_trimmed[type=="cycle_qual_counts",.(trimmed="untrimmed",median=weightedMedian(y,z)),by=c("sample","read","x")])
tplot[,read:=paste0(trimmed,".",read)]

bpin_info <- sampleoverview_raw[,c("sample","read","basepairs")]
bpin_info <- dcast(bpin_info,sample~read,value.var = "basepairs")
colnames(bpin_info) <- c("sample","in1_bp","in2_bp")

trim_summary<-readData(trim_summary.file, samplenames)
trim_summary<-trim_summary[bpin_info,on="sample"]
trim_summary[,dropped_reads:=in_reads-out_reads]
setnames(trim_summary,c("out_bp","w/adapters","w/adapters2"),c("out1_bp","adapter1reads","adapter2reads"))
trim_summary[,dropped1_bp:=in1_bp-out1_bp]
trim_summary[,dropped2_bp:=in2_bp-out2_bp]
tosetrelative.topairs <- c("out_reads","adapter1reads","adapter2reads","dropped_reads")
trim_summary[,(tosetrelative.topairs):=lapply(.SD,function(col)round(col/in_reads*100,2)),.SD=tosetrelative.topairs]

trim_summary.tomelt <- copy(trim_summary)
trim_summary.tomelt[,status:=NULL]
numericals <- colnames(trim_summary.tomelt)[sapply(trim_summary.tomelt,is.numeric)]
trim_summary.tomelt[,(numericals):=lapply(.SD,as.double),.SD=numericals]
trim_summary_molten <- melt(trim_summary.tomelt,id.vars = "sample")
```

Overview
==================

Row
--------------------------------------

### Application Note

The sequencing adapters should be removed from the reads, the sequencer can run its own adapter cutting, but as the reads aren't perfect and sometimes the adapter can appear within the sequence, depending on the preparation protocol used. 

Samples online in NGS read archives sometimes still have the adapters attached. The adaper sequence can be guessed by merging tools, as it will appear in the non overlapping region. If the reads don't contain the adapters this detection will produce long sequences with high percentage of Ns.

Quality trimming isn't performed at this point, as most assemblers are quality aware.

Whether the trimmed or untrimmed sequences are being used for the following workflow steps is configurable. To analyze the impact of adapter trimming, the merging step is done with both trimmed and untrimmed sequences.  

> Rendered at: `r Sys.time()`.

Row
--------------------------------------

### File Note

```{r}
knitr::kable(genInfoBlock(files))
```
> The table shows the files which have been used to generate this report.

Row
--------------------------------------

### BBmerge adapter detection

```{r}
trimmedadpaters<-merge_trimmed[,c("sample","adaptercount","Read1_adapter","Read2_adapter")]
untrimmedadpaters<-merge_untrimmed[,c("sample","adaptercount","Read1_adapter","Read2_adapter")]
adapters<-merge(untrimmedadpaters,trimmedadpaters,suffixes=c(".untrimmed",".trimmed"),by="sample")
adapters[,conf_Readadapter:=unlist(conf_adapters[sample])]
adaptercols <- colnames(adapters)[grepl("Read",colnames(adapters),fixed=T)]
adapters[,(adaptercols):=lapply(.SD,insertEvery,every=6),.SD=adaptercols]
setnames(adapters,c("sample", "adaptercount.untrimmed", "Read1_adapter.untrimmed", "Read2_adapter.untrimmed", "adaptercount.trimmed", "Read1_adapter.trimmed", "Read2_adapter.trimmed", "conf_Readadapter"),c("Sample","% Adapters detected\n(untrimmed)","R1 Adapter detected\n(untrimmed)","R2 adapter detected\n(untrimmed)","% Adapters detected\n(trimmed)","R1 Adapter detected\n(trimmed)","R2 adapter detected\n(trimmed)","Adapter specified in config"))
DT::datatable(adapters)
```

> This table shows the the adapter detected by [BBmerge](https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/bbmerge-guide/). The adapter sequences is detected by looking the non overlapping part of the merged reads. The adapter cutting is done with [cutadapt](https://cutadapt.readthedocs.io/en/stable/guide.html). 


Row
-----------------------------------------------------------------------

### Read pairs not written

```{r}
data = trim_summary_molten[variable %in% c("dropped_reads")]
ggplotly(ggplot(data, aes(x=sample, y=value,fill=sample)) + geom_bar(stat="identity") + xlab ("Sample") + ylab("Discarded Read pairs") + labs(fill = "Sample") + theme(legend.position = "none") + scale_y_continuous(labels = function(x) paste0(x, "%")))
```

>The number of read pairs discarded by cutadapt, if the plot is empty, no reads were discarded.

Row
------------------------------------

### Basepairs lost

```{r}
bps = trim_summary[,.SD,.SD=c("sample",colnames(trim_summary)[grepl("bp",colnames(trim_summary))])]
bps[,dropped1_bp:=as.double(dropped1_bp/in1_bp*1000)]
bps[,dropped2_bp:=as.double(dropped2_bp/in2_bp*1000)]
bps <- melt(bps[,.(sample,dropped1_bp,dropped2_bp)],id.vars = "sample")
bps[,Read:=gsub("dropped(\\d).+","R\\1",variable)]
bps[,variable:=gsub("read\\d(.+)","\\1",variable)]
bps = bps[,.(sample,Read,bpscut=value)]
ggplotly(ggplot(bps) + aes(x=sample, y=bpscut,fill=Read)  + geom_bar(stat="identity",position = "dodge") + xlab("Sample") + scale_y_continuous(labels = function(x) paste0(x, "â€°")) + ylab("Basepairs cut"))
```


>The number of basepairs cut by cutadapt in each sample.


### Read length change

```{r}
ggplot(trim_comparison) + aes(x=read,lower=quantile25ReadLenChange,upper=quantile75ReadLenChange,middle=medianReadLenChange,ymin=minReadLenChange,ymax=maxReadLenChange,group=interaction(sample,read),fill=sample) + geom_boxplot(stat="identity",position = "dodge")+ xlab ("Read group") +ylab("Read length in bp") + labs(fill="Sample") + theme(legend.position = "bottom")
```

> The length change of the reads. The whiskers extend to minimum and maximum value.

Row
--------------------------------------

### Change of the average quality for each read

```{r}
ggplot(trim_comparison) + aes(x=read,lower=quantile25AvgQualChange,upper=quantile75AvgQualChange,middle=medianAvgQualChange,ymin=minAvgQualChange,ymax=maxAvgQualChange,group=interaction(sample,read),fill=sample) + geom_boxplot(stat="identity",position = "dodge")+ xlab ("Read group") +ylab("Average Sequence Quality change") + labs(fill="Sample") + theme(legend.position = "bottom")
```


> The change of the average quality (mean of all phred scores for a sequence). The whiskers extend to minimum and maximum value.


### Median Cycle Quality Score

```{r}
ggplotly(ggplot(tplot) + aes(x=x,y=median,color=sample,linetype=read) +geom_line()+ ylab( "Median phred score")+xlab("Cycle") + theme(legend.title = element_blank()) )
```

> The median of the phred score at the given cycle/position for a sample and read direction

Row
--------------------------------------

### Trimmed files

```{r}
sampleoverview_trim.vis <- sampleoverview_trim[,.(sample,read,minReadLen,maxReadLen,reads)]
sampleoverview_trim.vis <- fastqcreports[sampleoverview_trim.vis,on=c("sample","read")]
colnam = c(sample="Sample",read="Read Direction",qcreport="FASTQC Report",minReadLen="Shortest Read [bp]",maxReadLen="Longest Read [bp]",reads="Number of read pairs")
setnames(sampleoverview_trim.vis,names(colnam),colnam)
datatable(sampleoverview_trim.vis,escape = F)
```

> The trimmed files. The links to the read files themself are not contained within the Analysis directory, but the [FASTQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) reports.
