---
title: "CAZy Read Annotation"
output: 
  flexdashboard::flex_dashboard:
    orientation: rows
    vertical_layout: scroll
    theme: sandstone
---

```{r, include=FALSE,message=FALSE}
if(exists("snakemake")){
  database.file <- snakemake@input[["db"]]
  hits.file = snakemake@input[["hits"]]
  commonscript.file = snakemake@input[["commons"]]
  readanno_sampled_overview.file = snakemake@input[["readanno_sampled_overview"]]
  threads=snakemake@threads
}else{
  database.file = "/run/user/2496/gvfs/sftp:host=highmem1/data1/2019_LJ_data/bachelor_thesis/metadiet/output/databases/readanno/prot/CAZyDB.tsv"
  hits.file = "/run/user/2496/gvfs/sftp:host=highmem1/data1/2019_LJ_data/bachelor_thesis/metadiet/output/Analysis/read_annotation/CAZyDB/CAZyDB_combineddata.tsv"
  readanno_sampled_overview.file = "/run/user/2496/gvfs/sftp:host=highmem1/data1/2019_LJ_data/bachelor_thesis/metadiet/output/Analysis/sequence_data/readanno_sampled/readanno_sampled_overview.tsv"
  commonscript.file = "commonReport.R"
  threads=NA
}
taxcols=c("superkingdom","phylum","class","order","family","genus","species")
files = c(hits.file,database.file)
source(commonscript.file)
commonOptions(threads)
require(tidyr)


readanno_sampled_overview = fread(readanno_sampled_overview.file)
hits <- fread(hits.file)
database <- fread(database.file)


hits = rbind(hits,cbind(sample="Database",database),fill=T)
hits[,(c("ECs","CazySubClasses","CazyClasses")):=extractCAZy(CAZyECs)]

cazydat = hits[sample!="Database",c("sample",taxcols,"CazySubClasses"),with=F]
cazydat = as.data.table(unnestSingleColUnique(cazydat,"CazySubClasses"))
cazydat[,CazyClasses:=mclapply(CazySubClasses,sub,pattern="(\\D{2,3}).+",replacement="\\1")]
cazydat[,CazyClasses:=sapply(CazyClasses,"[[",1)]
cazydat = cazydat[,.N,by=c("sample","CazyClasses","CazySubClasses",taxcols)]
```

Overview
=================

Row
--------------------------------------

### Application Note

The [carbohydrate active enzymes database(CAZy)](http://www.cazy.org/) classifies the enzymes according to their catalytic activity on carbohydrates and glycoconjugates and provides information on those classes. The [dbCAN project](http://bcb.unl.edu/dbCAN2/) maintains a database of sequences in the CAZy database and automatic annotation of those proteins. The raw fasta file has been used to build a database for [DIAMOND BLASTX](https://github.com/bbuchfink/diamond). The more sophiscated approach for CAZy annotation described used by dbCAN is used for annotating the bins later in the workflow.

> Rendered at: `r Sys.time()`.

Row
--------------------------------------

### File Note

```{r}
knitr::kable(genInfoBlock(files))
```
> The table shows the files which have been used to generate this report.


Row
--------------------------------------

### Distribution of the main classes

```{r}
dat=cazydat[,.(N=sum(N)),by=c("sample","superkingdom","CazyClasses")]
dat[,rel:=N/sum(N)*100,by="sample"]
ggplotly(ggplot(dat)+aes_string(x="sample",y="rel",fill="CazyClasses")+ geom_bar(stat="identity")+ xlab("Sample") + scale_y_continuous(labels = function(x) paste0(x, "%"))+ ylab("Percentage of classified Reads") + labs(fill="CAZy Class") + facet_grid(cols=vars(superkingdom))+theme(axis.text.x = element_text(angle = 33)))
```

> The distribution of the main classes among the samples and domains.

Row
--------------------------------------


```{r,results="asis"}
dat=filtermax(cazydat,groupcol=c("CazyClasses","sample"),filtercol="CazySubClasses",maxn=30,additionalsubcols=c(),additionalparentcols =c())
dat[,rel:=N/sum(N)*100,by=c("sample")]
dat=as.data.table(complete(dat,sample,nesting(CazyClasses,CazySubClasses),fill=list(N=0,rel=0)))
orderHeatMap(dat,"sample",field = "rel",grouping = "CazySubClasses")
dat=split(dat,dat$CazyClasses)
dat=lapply(dat,function(dat)orderHeatMap(dat,"CazySubClasses",field = "rel",grouping = "sample"))
plots=lapply(dat,function(dat)ggplot(dat)+aes_string(x="sample",y="CazySubClasses",fill="rel") +labs(fill="% of filtered\n Reads")+ geom_raster() + xlab("Sample") + ylab("CAZy Subclass")+theme(axis.text.x = element_text(angle = 33)))  #+facet_grid(cols=vars(superkingdom))
plots = plots[names(plots)!="Unclassified"]
for (i in seq_along(plots)) {
    plotname = names(plots)[[i]]
    drawplotasis(paste0(plotname),"The 30 most common subgroups in every sample are shown in this plot, others have been aggregated into \"Other\". The subclasses have been sorted by using Wards clustering with euclidian distance in every group and the samples on all subgroups combined.",ggplotly(plots[[i]]),576*1.75)
    if(i!=length(plots))
      drawrow()
  }
```


BLAST Results
===========================

Row
---------------------------------------

```{r,results="asis",fig.width=16}
drawBlastPlots(hits[sample!="Database",],readanno_sampled_overview)
```


Taxonomy
===========================

Row
---------------------------------------

```{r,results="asis"}
drawSampleTaxplots(cazydat[sample!="Database",.(N=sum(N)),by=c("sample",taxcols)])
```
